crates/shared/src/db_order_conversions.rs

```diff
@@ -1,11 +1,52 @@
+use anyhow::{anyhow, Result};
 use database::orders::{
-    BuyTokenDestination as DbBuyTokenDestination, OrderKind as DbOrderKind,
-    SellTokenSource as DbSellTokenSource, SigningScheme as DbSigningScheme,
+    BuyTokenDestination as DbBuyTokenDestination, FullOrder as FullOrderDb,
+    OrderKind as DbOrderKind, SellTokenSource as DbSellTokenSource,
+    SigningScheme as DbSigningScheme,
 };
+use ethcontract::H160;
 use model::{
+    interaction::InteractionData,
     order::{BuyTokenDestination, OrderKind, SellTokenSource},
     signature::SigningScheme,
 };
+use number_conversions::big_decimal_to_u256;
+
+pub fn extract_pre_interactions(order: &FullOrderDb) -> Result<Vec<InteractionData>> {
+    anyhow::ensure!(
+        order.pre_interactions_tos.len() == order.pre_interactions_values.len(),
+        "invalid pre_interactions length"
+    );
+    anyhow::ensure!(
+        order.pre_interactions_tos.len() == order.pre_interactions_data.len(),
+        "invalid pre_interactions length"
+    );
+    let mut pre_interactions = Vec::new();
+    for i in 0..order.pre_interactions_tos.len() {
+        pre_interactions.push(InteractionData {
+            to: H160(
+                order
+                    .pre_interactions_tos
+                    .get(i)
+                    .expect("invalid pre_interaction stored")
+                    .0,
```

Comment:
ubernit: I would use array indexing here:
```suggestion
                order.pre_interactions_tos[i].0,
```